src/ltdc.rs

```diff
@@ -0,0 +1,323 @@
+        let rcc = unsafe { &(*RCC::ptr()) };
+
+        // Screen constants
+        let total_width: u16 = config.h_sync + config.h_back_porch + config.active_width + config.h_front_porch - 1;
+        let total_height: u16 = config.v_sync + config.v_back_porch + config.active_height + config.v_front_porch - 1;
+        let lcd_clk: u32 = (total_width as u32) * (total_height as u32) * (config.frame_rate as u32);
+
+        // Enable LTDC peripheral's clock
+        rcc.apb2enr.modify(|_, w| w.ltdcen().enabled());
+        // Reset LTDC peripheral
+        rcc.apb2rstr.modify(|_, w| w.ltdcrst().reset());
+        rcc.apb2rstr.modify(|_, w| w.ltdcrst().clear_bit());
+
+        // Get base clock and PLLM divisor
+        let base_clk: u32;
+        match &hse {
+            Some(hse) => base_clk = hse.freq,
+            // If no HSE is provided, we use the HSI clock at 16 MHz
+            None => base_clk = 16_000_000
+        }
+        let pllm: u8 = rcc.pllcfgr.read().pllm().bits();
+
+        // There are 24 combinations possible for a divisor with PLLR and DIVR
+        // We find the one that is the closest possible to the target value
+        // while respecting all the conditions
+        let vco_in_mhz: f32 = (base_clk as f32 / pllm as f32) / 1_000_000.0;
+        let lcd_clk_mhz = (lcd_clk as f32) / 1_000_000.0;
+        let allowed_pllr = [2.0, 3.0, 4.0, 5.0, 6.0, 7.0];
+        let allowed_divr = [2.0, 4.0, 8.0, 16.0];
+        let mut best_pllr: f32 = allowed_pllr[0];
+        let mut best_divr: f32 = allowed_divr[0];
+        let mut best_plln: f32 = 100.0;
+        let mut best_error: f32 = (vco_in_mhz * best_plln) / (best_pllr * best_divr);
+        let mut error: f32;
+        let mut plln: f32;
+
+        for pllr in &allowed_pllr {
+            for divr in &allowed_divr {
+                plln = ((lcd_clk_mhz * divr * pllr) / vco_in_mhz).floor();
+                error = lcd_clk_mhz - (vco_in_mhz * plln) / (pllr * divr);
+                
+                // We have to make sure that the VCO_OUT is in range [100, 432]
+                // MHz Because VCO_IN is in range [1, 2] Mhz, the condition
+                // PLLN in range [50, 432] is automatically satisfied
+                if 100.0 <= vco_in_mhz * plln && vco_in_mhz * plln <= 432.0
+                    && error >= 0.0
+                    && error < best_error {
+                        best_pllr = *pllr;
+                        best_divr = *divr;
+                        best_plln = plln;
+                        best_error = error;
+                }
+            }   
+        }
+
+        let pllsaidivr: u8 = match best_divr as u16 {
+            2   => 0b00,
+            4   => 0b01,
+            8   => 0b10,
+            16  => 0b11,
+            _   => unreachable!()
+        };
+        
+        // // Write PPLSAI configuration
+        rcc.pllsaicfgr.write(|w| unsafe{
+            w   .pllsain().bits(best_plln as u16)
```

Comment:
Some weird spacing here... fyi you can run `cargo fmt` on this file if you want

Reply:
Allright, I'll do that

Reply:
Allright, I commit it